### PR TITLE
Add "Release tracking" analyzer to track 'analyzer releases' <-> 'shipped rules'

### PR DESCRIPTION
**Pull request by mavasani**
Release tracking analyzer enables third party analyzer packages to define analyzer releases with associated versions. Each release can track the following changes:
1. Additions: Set of new analyzer rules that shipped for the first time in this release.
2. Removals: Set of old analyzer rules that shipped in an earlier release, but are removed starting this release.
3. Changes: Set of existing analyzer rules that shipped in an earlier release, but one of the following attributes of the diagnostic changed in this release:
   1. Category of the diagnostic.
   2. Default severity of the diagnostic
   3. Enabled by default status.

This analyzer expects that the third party analyzer project has two additional files, namely `AnalyzerReleases.Unshipped.md` and `AnalyzerReleases.Shipped.md`, to track this release specific metadata.
1. `AnalyzerReleases.Unshipped.md`: Additional file for the upcoming or unshipped analyzer release. This file will start empty at the beginning of each release. While working on the upcoming release, it will track additions/removals/changes to analyzer rules in the repo.
1. `AnalyzerReleases.Shipped.md`: Additional file for shipped analyzer releases. This file only tracks analyzer rules for shipped releases. It cannot be changed while work is in progress for upcoming release. When a new analyzer package is released, a new release with shipped package version should be created in the shipped file, and all the entries from the unshipped file should be mvoed under the new shipped release.

Release tracking analyzer provides the diagnostics and code fixes to enable analyzer authors to maintain the shipped and unshipped files. Analyzer author only requires to perform the following two manual tasks:
1. Create these two files and pass mark them as additional files for their analyzer project (steps explained in next section), and
2. Move all the entries from unshipped file to the shipped file after each release.

Consider the following example:

1. `AnalyzerReleases.Shipped.md`:

```
## Release 1.0
Rule ID | Category | Severity | HelpLink (optional)
--------|----------|----------|--------------------
CA1000  |  Design  |  Warning | CA1000_Documentation_Link
CA2000  | Security |  Info    | CA2000_Documentation_Link
CA3000  |  Usage   | Disabled | CA3000_Documentation_Link

## Release 2.0
Rule ID | Category | Severity | HelpLink (optional)
--------|----------|----------|--------------------
CA2000  | Security | Disabled | CA2000_Documentation_Link
*REMOVED*CA3000  |  Usage   |  Disable | CA3000_Documentation_Link
CA4000  |  Design  |  Warning | CA4000_Documentation_Link
```

2. `AnalyzerReleases.Unshipped.md`:
```
Rule ID | Category | Severity | HelpLink (optional)
--------|----------|----------|--------------------
CA5000  | Security |  Warning |
CA6000  |  Design  |  Warning |
```

Analyzer author has shipped 2 analyzer releases:
1. Version 1.0 shipped three rules: CA1000, CA2000 and CA3000.
2. Version 2.0 changed the default severity of a shipped rule CA2000 from 'Warning' to 'Disabled'. It removed a shipped rule CA3000 and added a new rule CA4000.
3. Upcoming release will add 2 new rules CA5000 and CA6000.

When the next release is shipped, say version '3.0', a new release section for 3.0 should be created at the end of the shipped file and the entries from unshipped file should be moved under it. The files will change as below:

1. `AnalyzerReleases.Shipped.md`:

```
## Release 1.0
Rule ID | Category | Severity | HelpLink (optional)
--------|----------|----------|--------------------
CA1000  |  Design  |  Warning | CA1000_Documentation_Link
CA2000  | Security |  Info    | CA2000_Documentation_Link
CA3000  |  Usage   | Disabled | CA3000_Documentation_Link

## Release 2.0
Rule ID | Category | Severity | HelpLink (optional)
--------|----------|----------|--------------------
CA2000  | Security | Disabled | CA2000_Documentation_Link
*REMOVED*CA3000  |  Usage   |  Disable | CA3000_Documentation_Link
CA4000  |  Design  |  Warning | CA4000_Documentation_Link

## Release 3.0
Rule ID | Category | Severity | HelpLink (optional)
--------|----------|----------|--------------------
CA5000  | Security |  Warning |
CA6000  |  Design  |  Warning |
```

2. `AnalyzerReleases.Unshipped.md` (empty):
```
```

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @dotnet/dotnet-analyzers. This analyzer is the first step towards tooling and infrastructure in the repo to enable warning waves feature for .NET5 analyzers. This analyzer enables us to track which .NET analyzer rules ship with each shipped analyzer release and upcoming unshipped release.

Once this PR is merged in, I will enable this analyzer on the .NET analyzers project so we can start tracking the current rules in the project, with their category, default severity and enabled by default status.

Once that is done, I plan to write a simple post-build tool in this repo to convert the shipped and unshipped release tracking files into per-release .editorconfig files that are embedded into the analyzer NuGet package.

This analyzer NuGet package will be consumed by the SDK repo. The SDK repo tooling can choose the appropriate global .editorconfig file, based on user-facing MSBuild properties, and pass it into the compiler to enable the end-to-end warning wave support. NOTE: Compiler feature request https://github.com/dotnet/roslyn/issues/42219 needs to be implemented for the last step.

File: src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+Release tracking analyzer enables third party analyzer packages to define analyzer releases with associated versions. Each release can track the following changes:
+1. Additions: Set of new analyzer rules that shipped for the first time in this release.
+2. Removals: Set of old analyzer rules that shipped in an earlier release, but are removed starting this release.
+3. Changes: Set of existing analyzer rules that shipped in an earlier release, but one of the following attributes of the diagnostic changed in this release:
+   1. Category of the diagnostic.
+   2. Default severity of the diagnostic
+   3. Enabled by default status.
+
+This analyzer expects that the thirdparty analyzer project has two additional files, namely `AnalyzerReleases.Unshipped.md` and `AnalyzerReleases.Shipped.md`, to track this release specific metadata.
```
Comment (dotpaul) [inline src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md]: >the [](start = 27, length = 3)

Should mention that each analyzer project needs its own pair of files. If there multiple analyzer assemblies in a package, then they'll each need their own unique pair, right?

Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md]: Yes, that is correct.

File: src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+1. `AnalyzerReleases.Unshipped.md`: Additional file for the upcoming or unshipped analyzer release. This file will start empty at the beginning of each release. While working on the upcoming release, it will track additions/removals/changes to analyzer rules in the repo.
+1. `AnalyzerReleases.Shipped.md`: Additional file for shipped analyzer releases. This file only tracks analyzer rules for shipped releases. It cannot be changed while work is in progress for upcoming release. When a new analyzer package is released, a new release with shipped package version should be created in the shipped file, and all the entries from the unshipped file should be mvoed under the new shipped release.
```
Comment (dotpaul) [inline src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md]:
```suggestion
2. `AnalyzerReleases.Shipped.md`: Additional file for shipped analyzer releases. This file only tracks analyzer rules for shipped releases. It cannot be changed while work is in progress for upcoming release. When a new analyzer package is released, a new release with shipped package version should be created in the shipped file, and all the entries from the unshipped file should be mvoed under the new shipped release.
```

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer_ReleaseTracking.cs [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+        private static bool TryGetReleaseTrackingFiles(
+            ImmutableArray<AdditionalText> additionalTexts,
+            CancellationToken cancellationToken,
+            [NotNullWhen(returnValue: true)] out AdditionalText? shippedText,
+            [NotNullWhen(returnValue: true)] out AdditionalText? unshippedText)
+        {
+            shippedText = null;
+            unshippedText = null;
+
+            StringComparer comparer = StringComparer.Ordinal;
```
Comment (dotpaul) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer_ReleaseTracking.cs]: >Ordinal [](start = 53, length = 7)

OrdinalIgnoreCase? Or try Ordinal first, fallback to OrdinalIgnoreCase?

Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer_ReleaseTracking.cs]: Changed to `OrdinalIgnoreCase`, thanks.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer_ReleaseTracking.cs [hunk 4]
```diff
@@ -0,0 +1,900 @@
...
+            if (!string.IsNullOrEmpty(helpLink))
+            {
+                entry += $" [Documentation]({helpLink})";
+            }
+
+            return entry;
+
+            static string GetCategoryText(string? category)
+                => category ?? UndetectedText;
+
```
Comment (dotpaul) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer_ReleaseTracking.cs]: Totally an edge case, but are there any restrictions on category names not containing new lines and pipes?

Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer_ReleaseTracking.cs]: Hmm, I haven’t seen multiline categories in any package. We should consider those as unsupported for all practical purposes.

**Comment by mavasani** [comment]
Thanks a lot @dotpaul! I will soon turn this on for all packages in our repo in both 2.9.x and master branches.

**Pull request merged**